OCM-5576 | fix:  Edit autoscaling max replicas of nodepool force to s…

chenz4027 opened the pull request:
https://issues.redhat.com/browse/OCM-5576

marcolan018 commented on cmd/edit/machinepool/nodepool.go:
```diff
@@ -307,12 +307,25 @@ func editAutoscaling(nodePool *cmv1.NodePool, minReplicas int, maxReplicas int)
+	existingMaxReplica := nodePool.Autoscaling().MaxReplica()
+
+	if existingMinReplica != minReplicas {
+		if minReplicas >= 1 {
```
should `changed = true` only when `minReplicas >= 1`?

marcolan018 commented on cmd/edit/machinepool/nodepool.go:
```diff
@@ -307,12 +307,25 @@ func editAutoscaling(nodePool *cmv1.NodePool, minReplicas int, maxReplicas int)
-	if nodePool.Autoscaling().MaxReplica() != maxReplicas && maxReplicas >= 1 {
-		asBuilder = asBuilder.MaxReplica(maxReplicas)
+	if existingMaxReplica != maxReplicas {
+		if maxReplicas >= 1 {
```
same as above, `changed = true` should not be here if maxReplicas not set?

chenz4027 replied:
Thanks @marcolan018 , updated with the change

gdbranco commented on cmd/edit/machinepool/nodepool.go:
```diff
@@ -307,13 +307,26 @@ func editAutoscaling(nodePool *cmv1.NodePool, minReplicas int, maxReplicas int)
+			changed = true
+		}
+		if maxReplicas == 0 {
+			asBuilder = asBuilder.MaxReplica(existingMaxReplica)
```
Would it be ok to set it to the existing values initially instead to save on some condition branching?

chenz4027 replied:
Haha I actually did that first and refactored it to what it looks like now - will make the change

marcolan018 commented:
/lgtm

ziccardi commented on cmd/edit/machinepool/nodepool.go:
```diff
@@ -307,13 +307,27 @@ func editAutoscaling(nodePool *cmv1.NodePool, minReplicas int, maxReplicas int)
+	}
+
+	if existingMinReplica != minReplicas || existingMaxReplica != maxReplicas {
+		if min >= 1 && max >= 1 {
+			asBuilder = asBuilder.MinReplica(min).MaxReplica(max)
+			changed = true
+		}
```
```suggestion
		if min >= 1 && max >= 1 {
			return cmv1.NewNodePoolAutoscaling().MinReplica(min).MaxReplica(max)
		}
```

the `asBuilder` is returned only if the code enters this path. We can return it directly here and get rid of the `asBuilder` and `changed` variables declared at the start of the method.

chenz4027 replied:
Thanks @ziccardi , will make the change

chenz4027 commented:
@ziccardi I updated the refactor.

@gdbranco Provisioned a hcp and tested switching autoscaling on and off. 

```
./rosa edit machinepool workers -c magchen-hcp --enable-autoscaling --min-replicas 2 --max-replicas 6
I: Updated machine pool 'workers' on hosted cluster 'magchen-hcp'
./rosa edit machinepool workers -c magchen-hcp -i                        
? Enable autoscaling: No
? Replicas: 2
? Labels (optional): 
? Taints (optional): 
? Autorepair: Yes
? Node drain grace period (optional): 
I: Updated machine pool 'workers' on hosted cluster 'magchen-hcp'
```

Looks fine 
```
./rosa describe cluster -c magchen-hcp

Name:                       magchen-hcp
Domain Prefix:              magchen-hcp
Display Name:               magchen-hcp
...
Nodes:
 - Compute (desired):       2
 - Compute (current):       2

```

ziccardi commented:
/lgtm

ciaranRoche reviewed (approved):
/lgtm
